README.md
# ESP8266-WiFi-Socket-II

## Alternative Firmware for

- Shelly 1 OS, [https://shelly.cloud/shelly1-open-source/](https://shelly.cloud/shelly1-open-source/)

- BlitzWolf SHP6, [https://www.blitzwolf.com/2300W-EU-WIFI-Smart-Socket-p-300.html](https://www.blitzwolf.com/2300W-EU-WIFI-Smart-Socket-p-300.html)

- OBI Version 2, [https://www.obi.de/hausfunksteuerung/wifi-stecker-schuko-weiss/p/4077806](https://www.obi.de/hausfunksteuerung/wifi-stecker-schuko-weiss/p/4077806)

- OBI Version 1, [https://www.obi.de/hausfunksteuerung/wifi-stecker-schuko/p/2291706](https://www.obi.de/hausfunksteuerung/wifi-stecker-schuko/p/2291706)

## Precompiled Firmware

### Attention - update OTA from 1.1.x to 2.x for OBI Version 1 only

> If you like to update from firmware 1.1.x (OBI Version 1) to 2.x over the air (OTA) please first flash firmware version 1.2.0migration.
>
> If you try to update direct from 1.1.x to the current 2.x firmware
> version you will get an "out of memory" error
>
> Due to a new memory model the firmware will reset all settings to the default values

Find the precompiled firmware in the [firmware](https://github.com/thorsten-l/ESP8266-WiFi-Socket-II/tree/master/firmware) directory.

## Compile

This is a [PlatformIO](https://platformio.org/) project.

## Firmware Reset

1. Power on the WiFi Socket (plug in)
2. Press down and hold the Power-Button during the first 5s
3. Release Power-Button when the WiFi LED starts rapidly fast blinking
4. The firmware reset ends after about 30s indicated by a permanent WiFi LED on.

## Default Settings

After a firmware reset the following values are set.

|                            | value                 | description                           |
| -------------------------- | --------------------- | ------------------------------------- |
| Admin password             | admin                 | Setup admin password                  |
| WiFi Mode                  | AP                    | Accesspoint- or Station-Mode          |
| WiFi SSID                  | WifiSocketII-`esp id` | SSID in accesspoint mode              |
| WiFi Password              | 12345678              | Default WiFi password                 |
| Network Mode               | DHCP                  | Network mode STATIC or DHCP \*1       |
| Network IP-Address         | 192.168.192.1         | device ip address \*2                 |
| Network Mask               | 255.255.255.0         | Subnet mask \*2                       |
| Network Gateway            | 192.168.192.1         | Default gateway \*2                   |
| Network DNS                | 192.168.192.1         | DNS server address \*2                |
| OTA Hostname               | wifi-socket-1         | *O*ver *T*he *A*ir hostname           |
| OTA Password               | otapass               | OTA firmware upload password          |
| OpenHAB Callback Enabled   | **false**             |                                       |
| OpenHAB Version            | 1                     | Callback for OpenHAB version 1 or 2   |
| OpenHAB Item Voltage       | PowerMeter1Voltage    | ('-' = disables this item)            |
| OpenHAB Item Current       | PowerMeter1Current    | ('-' = disables this item)            |
| OpenHAB Item Power         | PowerMeter1Power      | ('-' = disables this item)            |
| OpenHAB Item Name          | WifiSocket1           |                                       |
| OpenHAB Host               | 192.168.1.1           |                                       |
| OpenHAB Port               | 80                    |                                       |
| OpenHAB Use Authentication | false                 |                                       |
| OpenHAB User               | user                  |                                       |
| OpenHAB Password           | password              |                                       |
| OpenHAB SendingInterval    | 60                    | interval time in seconds (0=disable)  |
| Alexa enabled              | **false**             |                                       |
| Alexa devicename           | Socket                | Devicename for Alexa HA               |
| MQTT Enabled               | **false**             |                                       |
| MQTT ClientId              | WifiSocket1           |                                       |
| MQTT Host                  | 192.168.1.1           |                                       |
| MQTT Port                  | 1883                  |                                       |
| MQTT Use Authentication    | false                 |                                       |
| MQTT User                  | user                  |                                       |
| MQTT Password              | password              |                                       |
| MQTT In Topic              | socket1/in            |  receiving commands                   |
| MQTT Out Topic             | socket1/out           |  sending callback                     |
| MQTT Topic Voltage         | powermeter1/voltage   |  ('-' = disables this topic)          |
| MQTT Topic Current         | powermeter1/current   |  ('-' = disables this topic)          |
| MQTT Topic Power           | powermeter1/power     |  ('-' = disables this topic)          |
| MQTT Topic JSON            | powermeter1/json      |  sending all values in JSON format    |
| MQTT Sending Interval      | 60                    |  interval time in seconds (0=disable) |
| Syslog Enabled             | **false**             |                                       |
| Syslog Host                | 192.168.1.1           |                                       |
| Syslog Port (TCP)          | 514                   |                                       |
| Syslog App Name            | gateway-doorbell      |                                       |

\*1 In WiFi Station-Mode only

\*2 will be overwritten from DHCP-Response

## Hardware

- [BlitzWolf SHP6](https://www.blitzwolf.com/2300W-EU-WIFI-Smart-Socket-p-300.html)

- [OBI Wifi Stecker Schuko - Version 2](https://www.obi.de/hausfunksteuerung/wifi-stecker-schuko-weiss/p/4077806)

- [OBI Wifi Stecker Schuko - Version 1](https://www.obi.de/hausfunksteuerung/wifi-stecker-schuko/p/2291706)

## OpenHAB

### OpenHAB - HTTP Binding sample

```t
Switch LivingRoom_Power "Living Room Outlet" {http=">[ON:GET:http://wifi-socket.ip/on] >[OFF:GET:http://wifi-socket.ip/off]"}
```

If you like to read the state of the socket by OpenHAB. First create a file `getState.js` in the `transform` directory with the following content:

```javascript
JSON.parse(input).state;
```

add

```t
<[http://wifi-socket.ip/state:30000:JS(getState.js)]
```

to the `Switch` definition. Every 30s the state of the socket will be checked.

### OpenHAB - MQTT binding sample for the switch and the energy monitor values

#### MQTT Binding Version 1.x

```t
Switch WifiSocketBwShp6 "BW SHP6 Ein / Aus" {mqtt=">[mosquitto:socket-bw/in:command:ON:1],>[mosquitto:socket-bw/in:command:OFF:0],<[mosquitto:socket-bw/out:state:default]"}

Number WifiSocketBwShp6Voltage "BW SHP6 Voltage [%.1fV]" {mqtt="<[mosquitto:socket-bw/voltage:state:default]"}

Number WifiSocketBwShp6Current "BW SHP6 Current [%.2fA]" {mqtt="<[mosquitto:socket-bw/current:state:default]"}

Number WifiSocketBwShp6Power "BW SHP6 Power [%.1fW]" {mqtt="<[mosquitto:socket-bw/power:state:default]"}
```

#### MQTT Binding Version 2.x

For the Binding Version 2.x it is necessary to configure the MQTT Thing and according to that the items.

mqtt.things:

```t
Bridge mqtt:broker:mosquitto "Mosquitto" [ host="127.0.0.1", secure=false, clientID="MotherShip", username="$MQTT_USERNAME", password="$MQTT_PASSWORD"]
{
   Thing topic socket-bw "BW SHP6" {
    Channels:
     Type switch : state "BW SHP6 Ein / Aus" [ stateTopic="wsocket-bw/out", on="ON", off="OFF", commandTopic="socket-bw/in", on="ON", off="OFF" ]
     Type string : voltage [ stateTopic="wsocket-bw/voltage" ]
     Type string : current [ stateTopic="wsocket-bw/current" ]
     Type string : power [ stateTopic="wsocket-bw/power" ]
   }

}
```

mqtt.items:

```t
Switch WifiSocketBwShp6 "BW SHP6 Ein / Aus" { channel="mqtt:topic:mosquitto:socket-bw:state" }

Number WifiSocketBwShp6Voltage "BW SHP6 Voltage [%.1fV]" { channel="mqtt:topic:mosquitto:socket-bw:voltage" }

Number WifiSocketBwShp6Current "BW SHP6 Current [%.2fA]" { channel="mqtt:topic:mosquitto:socket-bw:current" }

Number WifiSocketBwShp6Power "BW SHP6 Power [%.1fW]" { channel="mqtt:topic:mosquitto:socket-bw:power" }
```

## References

- [PlatformIO](https://platformio.org/)
- [Arduino core for ESP8266 WiFi chip](https://github.com/esp8266/Arduino)
- [Async TCP Library for ESP8266](https://github.com/me-no-dev/ESPAsyncTCP)
- [Async Web Server for ESP8266 and ESP32](https://github.com/me-no-dev/ESPAsyncWebServer)
- [FauxmoESP - Amazon Alexa support for ESP8266 and ESP32 devices](https://bitbucket.org/xoseperez/fauxmoesp)
- [Arduino-pubsubclient - A client library for the ESP8266 that provides support for MQTT](https://github.com/knolleary/pubsubclient)
- [Pure.CSS - A nice CSS, as GZIP it is less than 4k so it is useful for embedded devices](https://purecss.io/)


CHANGES.md
# CHANGES

## release 2.8.1

- root page state is more responsive

## release 2.8.0

- captive portal implementation
- captive portal DNS server for AP mode
- OpenHAB callback performance improvements
- check for invalid firmware filename implemented
- copy firmware.bin to firmware-version.pioenv.bin
- refresh to root page after 30s
- implementation for Shelly 1 OS switch

## release 2.7.0

- simple favicon added
- reorder openhab settings
- show millis as comment in setup page
- relpace all inline strings with define macros
- detailed WiFi info added to /state restful info
- new functions for backup and restore
- remove maintenance from setup page, now it has an own page
- Using JSON file format to store config

## release 2.6.0

- Implementation for BlitzWolf SHP6 inclusive energy monitor
- platform updated to version espressif8266@2.2.0
- own board definitions
  
## release 2.5.0

- reset firmware to system defaults added in setup page
- paramBool function moved to WebUtils

## release 2.4.2

- show chipinfo added
- new dns setting
- platform version 2.0.4
- initial travis configuration
- additional vscode exceptions

## release 2.4.1

- Pin initialization for development version was missing

## release 2.4.0

- html redesign of the setup page

## release 2.3.1

- show PIOENV_NAME in firmware upload label

## release 2.3.0

- update firmware over HTTP in setup page
- optimized platformio file structure
- web pages (handler) are now in seperate files
- added PIOENV_NAME to every platformio env
- show pioenv_name, hostname, spiffs_total and spiffs_used in info json

## release 2.2.2

- debug code removed
- remove unused code

## release 2.2.1

- using CPP macro for WIFI_LED ON / OFF
- complete rewriting of the setup page code
- dev1lite added

## release 2.1.0

- remove development firmware
- easier wifi network seletion, firmware is now available for both obi socket versions
- more information in info call
- show free heap in /info restful call

## release 2.0.6

- syslog messages added
- syslog enabled added to info page
- network and syslog added printConfig
- flashing esp info
- new screenshots for docs
- wrong default mqtt port in readme
- wrong default syslog port in readme

## release 2.0.5

- code modifications for fauxmoesp lib 3.1.0 (alexa)
- static network config and syslog config added


lib/App/DefaultAppConfig.h
#ifndef __DEFAULT_APP_CONFIG_H__
#define __DEFAULT_APP_CONFIG_H__

#include <ESP8266WiFi.h>

#define DEFAULT_WIFI_SSID "WifiSocketII-%06x"
#define DEFAULT_WIFI_PASSWORD "12345678"
#define DEFAULT_WIFI_MODE WIFI_AP

#define DEFAULT_NET_MODE NET_MODE_DHCP
#define DEFAULT_NET_HOST "192.168.192.1"
#define DEFAULT_NET_MASK "255.255.255.0"
#define DEFAULT_NET_GATEWAY "192.168.192.1"
#define DEFAULT_NET_DNS "192.168.192.1"

#define DEFAULT_OTA_HOSTNAME "wifi-socket-1"
#define DEFAULT_OTA_PASSWORD "otapass"

#define DEFAULT_ADMIN_PASSWORD "admin"

#define DEFAULT_OHAB_ENABLED false
#define DEFAULT_OHAB_VERSION 1
#define DEFAULT_OHAB_HOST "192.168.1.1"
#define DEFAULT_OHAB_PORT 80
#define DEFAULT_OHAB_USEAUTH true
#define DEFAULT_OHAB_USER "user"
#define DEFAULT_OHAB_PASSWORD "password"
#define DEFAULT_OHAB_ITEMNAME "WifiSocket1"
#define DEFAULT_OHAB_ITEM_VOLTAGE "WifiSocket1Voltage"
#define DEFAULT_OHAB_ITEM_CURRENT "WifiSocket1Current"
#define DEFAULT_OHAB_ITEM_POWER "WifiSocket1Power"
#define DEFAULT_OHAB_SENDING_INTERVAL 60

#define DEFAULT_ALEXA_ENABLED false
#define DEFAULT_ALEXA_DEVICENAME "Socket"

#define DEFAULT_MQTT_ENABLED false
#define DEFAULT_MQTT_CLIENTID "WifiSocket1"
#define DEFAULT_MQTT_HOST "192.168.1.1"
#define DEFAULT_MQTT_PORT 1883
#define DEFAULT_MQTT_USEAUTH true
#define DEFAULT_MQTT_USER "user"
#define DEFAULT_MQTT_PASSWORD "password"
#define DEFAULT_MQTT_INTOPIC "socket1/in"
#define DEFAULT_MQTT_OUTTOPIC "socket1/out"
#define DEFAULT_MQTT_TOPIC_VOLTAGE "socket1/voltage"
#define DEFAULT_MQTT_TOPIC_CURRENT "socket1/current"
#define DEFAULT_MQTT_TOPIC_POWER "socket1/power"
#define DEFAULT_MQTT_TOPIC_JSON "socket1/json"
#define DEFAULT_MQTT_SENDING_INTERVAL 60

#define DEFAULT_SYSLOG_ENABLED false
#define DEFAULT_SYSLOG_HOST "192.168.1.1"
#define DEFAULT_SYSLOG_PORT 514
#define DEFAULT_SYSLOG_APP_NAME "wifi-socket"

#define DEFAULT_POWER_BUTTON_MODE POWER_BUTTON_MODE_TOGGLE_SWITCH

#define DEFAULT_LED_NIGHT_MODE_ENABLED false
#define DEFAULT_LED_NIGHT_MODE_TIMEOUT 15

#endif
